### PR TITLE
Require C11 _Atomics

### LINKED ISSUE #485: Move to C11 and adopt _Atomic, static_asserts, and thread_local
As per https://github.com/valkey-io/valkey/commit/445a4b669a3a7232a18bf23340c5f7d580aa92c7, we implemented custom logic to compile with C11 when atomics were available, and C99 when they were not available. However, with the EoL of CentOS 7, we believe that all default compilers should be able to compile with C99. This gives us access to three pieces of functionality we use:
1. Static asserts, we have some custom code defined to support it but now we can standardize on the C11 functionality. 
2. native thread_local support. We can avoid `_thread`.
3. Native atomics support, so we can remove our custom logic supporting atomics.

This change should target Valkey 8. We will continue to use the old gcc version on Valkey 7.2.

### PR DESCRIPTION
**Pull request by adetunjii**
- Replaces custom atomics logic with C11 default atomics logic.
- Drops  "atomicvar_api" field from server info

Closes #485

### CODE REVIEW

File: src/module.c
Comment (aiven-sal) [inline src/module.c]: It looks like you forgot to replace this `atomicGet` with `atomic_load_explicit` here.

**Review by zuiderkwast — COMMENTED** [review]
Looks good so far.

I think we can also wait for another changes which is in progress. It's is a change for I/O threads which will probably use atomic variables a lot.

My idea is that we replace all atomic{Get,Set,Incr} and also delete atomicvar.h in one PR just before Valkey 8 release candidate 1, when no new features will be added. Then, if some users have a problem when testing the release candidate, we can revert it before we release 8.0.

@madolson What do you think?

**Comment by madolson** [comment]
@zuiderkwast I guess i would prefer we do the C11 migration now. Then anyone who is building from unstable will be able to start relying on the new improvements. I would rather find out sooner that someone might still have build issues.

File: src/bio.c
Comment (aiven-sal) [inline src/bio.c:265]: You correctly rewrote the old code, but, to me, it looks like the old code itself can be improved.
Setting _status and _errno in this way, means that, when those values are read in https://github.com/valkey-io/valkey/pull/490/files#diff-1abc5651133d108c0c420d9411925373c711133e7748d9e4f4c97d5fb543fdd9R4591-R4593 ,  _errno may be not set yet.
Now, maybe it isn't very important to preserve the correct errno in this case, but considering how rare (hopefully) this kind of failures are, it shouldn't hurt much to move to a less relaxed memory ordering.
Maybe some form of Release/Acquire.
E.g.
in bio.c :
```
        atomic_store_explicit(&server.aof_bio_fsync_errno,errno,memory_order_relaxed);
        atomic_store_explicit(&server.aof_bio_fsync_status,C_ERR,memory_order_release);
```
in server.c:
```
        int aof_bio_fsync_status = atomic_load_explicit(&server.aof_bio_fsync_status,memory_order_relaxed);
        if (aof_bio_fsync_status == C_ERR) {
            atomic_thread_fence(memory_order_acquire)
            server.aof_last_write_errno = atomic_load_explicit(&server.aof_bio_fsync_errno,memory_order_relaxed);
            return DISK_ERROR_TYPE_AOF;
        }
```
This should perform just like the relaxed code when no error occurs and it will incur some small performance degradation only if fsync fails.

What do you think?

**Review by aiven-sal — COMMENTED** [review]
You need to remove https://github.com/valkey-io/valkey/blob/dd2b27ea4eeecf0b02d3b985fff7de3c49dc1e14/src/zmalloc.c#L54
and replace https://github.com/valkey-io/valkey/blob/dd2b27ea4eeecf0b02d3b985fff7de3c49dc1e14/src/server.c#L5631 with something meaningful (I'm not sure what that would be). 
And then you can remove atomicvar.h completely.

**Comment by zuiderkwast** [comment]
> and replace
> 
> https://github.com/valkey-io/valkey/blob/dd2b27ea4eeecf0b02d3b985fff7de3c49dc1e14/src/server.c#L5631
> with something meaningful (I'm not sure what that would be).

"c11-builtin" is what's used today when C11 atomics are used.

We can also consider dropping this info field.

**Comment by adetunjii** [comment]
> > and replace
> > https://github.com/valkey-io/valkey/blob/dd2b27ea4eeecf0b02d3b985fff7de3c49dc1e14/src/server.c#L5631
> > 
> > with something meaningful (I'm not sure what that would be).
> 
> "c11-builtin" is what's used today when C11 atomics are used.
> 
> We can also consider dropping this info field.

@zuiderkwast I think we should just drop it since we're making C11 the standard.

**Comment by zuiderkwast** [comment]
Ok, please mention it in the PR description that we're dropping the INFO field "atomicvar_api". It's a user-facing change.

File: src/bio.c
Comment (adetunjii) [reply src/bio.c:265]: The performance degradation should be negligible.

File: src/server.c
Comment (aiven-sal) [inline src/server.c]: My suggestion was to put the fence inside the `if` block, because it's only needed if you are going to access _errno. It that way you incur no performance penalty when _status != C_ERR (not that this is necessarily important).
If you think that the code would be clearer and more maintainable applying the fence unconditionally (I can agree with this), you can just use `memory_order_acquire` directly in `atomic_load_explicit` instead of `memory_order_relaxed` and drop the `atomic_thread_fence` call.

**Review by aiven-sal — COMMENTED** [review]
I think you can now `rm src/atomicvar.h` .
Thanks for taking the time to work on my comments!

File: src/server.c
Comment (adetunjii) [reply src/server.c]: For clarity sake, the best would be to use memory_order_acquire directly in atomic_load_explicit.

**Review by aiven-sal — APPROVED** [review]
LGTM
Thank you!

**Comment by zuiderkwast** [comment]
We need to remove all CI jobs for CentOS 7. They fail because it doesn't support C11. Shall we change these jobs to CentOS 8? I think the point is to have one of the oldest OSes that's still supported. (CentOS 7 is EOL in June this year.)

**Comment by adetunjii** [comment]
> We need to remove all CI jobs for CentOS 7. They fail because it doesn't support C11. Shall we change these jobs to CentOS 8? I think the point is to have one of the oldest OSes that's still supported. (CentOS 7 is EOL in June this year.)

Yeah. I think it's best to remove old CI jobs for CentoS 7. Let's fix that in another PR

**Comment by zuiderkwast** [comment]
> I think it's best to remove old CI jobs for CentoS 7. Let's fix that in another PR

I don't like to merge a PR which breaks the CI. It affects the daily CI too. We can do it in a separate PR, but then we should merge that CI PR before we merge this PR.

**Comment by adetunjii** [comment]
Okay sure. I’ll have a look at it

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by madolson** [comment]
> Shall we change these jobs to CentOS 8?

CentOS 8 is already EoL, it never got long term support from RHEL from the looks of it. CentOS also doesn't exist anymore, it was evolved into CentOS stream. We could try AlmaLinux 9 or CentOS stream 9 as a replacement. I pinged some folks, but my initial guess is to move to CentOS stream. 

EDIT: Ideally we should have a matrix and do CentOS stream 9 and AlmaLinux 8. AlmaLinux 8 will be maintained until 2029, so it seems like a great replacement for our CentOS 7 test.

File: src/networking.c
Comment (zuiderkwast) [inline src/networking.c:4152]: What's defined for this in atomicvar.h is

    atomic_load_explicit(&var, memory_order_seq_cst)

This or memory_order_acquire, which is equivalent for a load operation.

File: src/networking.c
Comment (zuiderkwast) [inline src/networking.c:4157]: atomic_store_explicit with release or seq_cst is the equivalent of atomicSetWithSync.

**Review by zuiderkwast — COMMENTED** [review]
I've looked though this diff now. Looks good, but I think I spotted two typos.

File: src/networking.c
Comment (aiven-sal) [reply src/networking.c:4152]: I believe that `atomic_load(&var)` is equivalent to `atomic_load_explicit(&var, memory_order_seq_cst)`.
But it makes sense to be more explicit for clarity.

Comment (zuiderkwast) [reply src/networking.c:4152]: Oh, sorry! I didn't know. `atomic_load()` and `atomic_store()` are fine then.

File: src/networking.c
Comment (zuiderkwast) [reply src/networking.c:4157]: I realize atomic_store() is seq_cst, sorry!

**Review by zuiderkwast — APPROVED** [review]
LGTM.

Don't merge until we have replaced CentOS 7 in CI jobs.

**Comment by madolson** [comment]
I documented what we should do about centos 7 here: https://github.com/valkey-io/valkey/issues/527, in case someone else has time to pick it up.

**Comment by zuiderkwast** [comment]
Centos7 is still in the CI in this branch. I don't know how it happened. Did you merge latest unstable?

You can use merge commits. No need to rebase and force-push. We'll squash-merge it anyway.

**Comment by adetunjii** [comment]
> Centos7 is still in the CI in this branch. I don't know how it happened. Did you merge latest unstable?
> 
> You can use merge commits. No need to rebase and force-push. We'll squash-merge it anyway.

It should be fine now

**Comment by adetunjii** [comment]
@zuiderkwast when can this be merged?

**Comment by zuiderkwast** [comment]
> @zuiderkwast when can this be merged?

Probably tomorrow. People are not working today.

**Comment by zuiderkwast** [comment]
Actually we discussed it now and it seems fine to merge. :rocket:

**Pull request merged**
